### PR TITLE
Fix bad segfault happening at complex async DB read/write access

### PR DESCRIPTION
**Pull request by LefterisJP**
What was happening is the following:

1. Get in the progress callback but before the `in_callback` lock is acquired.
2. Wait for the lock acquisition and context switch out to other greenlets. .A lot of things happen and the callback gets entered by a lot of other sections and has the lock acquired. Some of those places do `connection.commit()` and modify the connection. As per the docs that's a no no: https://www.sqlite.org/c3ref/progress_handler.html
4. We context switch back to the original callback acquire the lock
and then release it and the callback is exited. Since the connection has been modified **KABOOM SEGFAULT**.

The way around this is:

1.To also acquire the callback lock before connection
modifications. So before commit and rollback.
2. If we get in the callback and see that the lock is acquired we do
not wait for it. Instead we exit the callback immediately. This
guarantees that in the case this happens we will not context switch
and exit the callback with the connection having been modified. We
just exit immediately so no modifications can ever happen so long as
we are in the callback.

This PR also adds a test that fails with the segfault if the fix is not committed and also adds a small bug fix for allowing setting the `sql_vm_instructions_cb` to 0

### CODE REVIEW

File: rotkehlchen/tests/db/test_async.py [hunk 1]
```diff
@@ -70,3 +97,65 @@ def test_async_segfault(database, function_scope_messages_aggregator):
...
+    gevent.joinall([a, b, c, d, e])
+
+
+@pytest.mark.parametrize('sql_vm_instructions_cb', [0])
+def test_can_disable_callback(database):  # pylint: disable=unused-argument
+    """Simply test that setting sql_vm_instructions_cb to 0 works
+
+    This is a regression test since setting to 0 was hitting an assertion before
+    """
+    assert True  # no need to do anything. Test would fail at fixture setup
```
Comment (yabirgb) [inline rotkehlchen/tests/db/test_async.py:161]: why not use pass instead of asserting true?

File: rotkehlchen/tests/db/test_async.py [hunk 2]
```diff
@@ -45,8 +70,10 @@ def read_actions(database, limit, msg_aggregator):
         )
 
 
-def test_async_segfault(database, function_scope_messages_aggregator):
+@pytest.mark.parametrize('sql_vm_instructions_cb', [100])
+def test_callback_segfault_simple(database, function_scope_messages_aggregator):
     """Test that the async and sqlite progress handler segfault yielding bug does not hit us
+    This one is protected against by having the lock inside the callback.
```
Comment (yabirgb) [inline rotkehlchen/tests/db/test_async.py:76]: is missing a word after against?

**Review by yabirgb — COMMENTED** [review]
Looks okey, two questions

File: rotkehlchen/tests/db/test_async.py (hunk 1 above)
Comment (LefterisJP) [reply rotkehlchen/tests/db/test_async.py:161]: mypy and pylint complained for useless pass

File: rotkehlchen/tests/db/test_async.py (hunk 2 above)
Comment (LefterisJP) [reply rotkehlchen/tests/db/test_async.py:76]: no

**Comment by LefterisJP** [comment]
All beautifully green. Merging.

**Pull request merged**